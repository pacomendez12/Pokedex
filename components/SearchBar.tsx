import i18n from "i18n-js";
import * as React from "react";
import {
  StyleSheet,
  Button,
  Platform,
  NativeSyntheticEvent,
  NativeTouchEvent,
} from "react-native";

import { Input, InputProps, View } from "../components/Themed";
import Colors from "../constants/Colors";

const isAndroid = Platform.OS === "android";

const SECONDARY_BASE_COLOR = "#8f8f91";

export default function SearchBar({ onSearch }: Props) {
  const [text, setText] = React.useState("");
  const [hasFocus, setHasFocus] = React.useState(false);
  const input = React.createRef<InputProps>();

  const shouldShowCancelButton = () => {
    return hasFocus || text !== "";
  };

  const handleFocus = () => {
    setHasFocus(true);
  };

  const handleBlur = () => {
    setHasFocus(false);
  };

  const renderCancelButton = () => {
    return (
      shouldShowCancelButton() && (
        <CancelButton
          onCancel={() => {
            input.current?.blur();
            setText("");
            onSearch("");
          }}
        />
      )
    );
  };

  const inputRightElement = shouldShowCancelButton()
    ? false
    : {
        type: "ionicon",
        name: "ios-mic",
        color: SECONDARY_BASE_COLOR,
      };

  return (
    <View
      style={styles.container}
      lightColor={Colors.light.secondaryBackground}
      darkColor={Colors.dark.background}
    >
      <Input
        ref={input}
        placeholder={i18n.t("Search")}
        containerStyle={styles.inputMainContainer}
        inputContainerStyle={styles.inputContainer}
        leftIcon={{
          type: "ionicon",
          name: "ios-search",
          color: SECONDARY_BASE_COLOR,
        }}
        rightIcon={inputRightElement}
        value={text}
        onChangeText={(value) => setText(value)}
        onFocus={handleFocus}
        onBlur={handleBlur}
        onSubmitEditing={() => onSearch(text)}
      />
      {renderCancelButton()}
    </View>
  );
}

const CancelButton = ({
  onCancel,
}: {
  onCancel: (ev: NativeSyntheticEvent<NativeTouchEvent>) => void;
}) => {
  return (
    <View
      style={{
        justifyContent: "center",
        paddingRight: 10,
      }}
      lightColor={Colors.light.secondaryBackground}
      darkColor={Colors.dark.background}
    >
      <Button
        color={isAndroid && Colors.shared.appColor}
        title={i18n.t("Cancel")}
        onPress={onCancel}
      />
    </View>
  );
};

type Props = {
  onSearch: (value: string) => void;
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    paddingTop: 10,
    paddingBottom: 10,
    justifyContent: "center",
    height: 59,
  },
  inputMainContainer: {
    flexGrow: 1,
    width: 100,
  },
  inputContainer: {
    height: 40,
    borderRadius: 10,
    paddingLeft: 10,
    paddingRight: 10,
    borderBottomWidth: 0,
  },
});
