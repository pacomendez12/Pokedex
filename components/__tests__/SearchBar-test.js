import { render, fireEvent, cleanup } from "@testing-library/react-native";
import * as Localization from "expo-localization";
import i18n from "i18n-js";
import * as React from "react";
import renderer from "react-test-renderer";

import SearchBar from "../SearchBar";

i18n.translations = {
  en: require("../../translations/en.json"),
  es: require("../../translations/es.json"),
};

i18n.locale = Localization.locale;
i18n.fallbacks = true;

describe("<SearchBar> tests", () => {
  afterEach(cleanup);

  it("searchBar by default shows only input component", () => {
    const tree = renderer.create(<SearchBar onSearch={() => {}} />).toJSON();
    expect(tree.children.length).toBe(1);
  });

  it("searchBar should show cancel button on focus", async () => {
    const mockFn = jest.fn();
    const { getByText, findByPlaceholderText } = render(
      <SearchBar onSearch={mockFn} />
    );

    const input = await findByPlaceholderText("Search");
    fireEvent(input, "focus");

    expect(getByText("Cancel")).not.toBeNull();
  });

  it("searchBar should not keep cancel button on blur if it doesn't contains text", async () => {
    const mockFn = jest.fn();
    const { findByPlaceholderText, toJSON } = render(
      <SearchBar onSearch={mockFn} />
    );

    const input = await findByPlaceholderText("Search");
    fireEvent(input, "focus");
    fireEvent(input, "blur");

    const tree = toJSON();
    expect(tree.children.length).toBe(1);
  });

  it("searchBar should keep cancel button on blur if contains text", async () => {
    const mockFn = jest.fn();
    const { getByText, findByPlaceholderText } = render(
      <SearchBar onSearch={mockFn} />
    );

    const input = await findByPlaceholderText("Search");
    fireEvent(input, "focus");
    fireEvent(input, "onChangeText", "something");
    fireEvent(input, "blur");

    expect(getByText("Cancel")).not.toBeNull();
  });

  it("searchBar call parent onSearch function with searchBar text", async () => {
    const mockFn = jest.fn();
    const { findByPlaceholderText } = render(<SearchBar onSearch={mockFn} />);

    const input = await findByPlaceholderText("Search");
    fireEvent(input, "focus");
    fireEvent(input, "onChangeText", "something");
    fireEvent(input, "onSubmitEditing");

    expect(mockFn).toBeCalledWith("something");
  });

  it("searchBar calls onSearch and later clear that value", async () => {
    const mockFn = jest.fn();
    const { getByText, findByPlaceholderText } = render(
      <SearchBar onSearch={mockFn} />
    );

    const input = await findByPlaceholderText("Search");
    fireEvent(input, "focus");
    fireEvent(input, "onChangeText", "something");
    fireEvent(input, "onSubmitEditing");

    fireEvent(input, "focus");

    fireEvent(getByText("Cancel"), "press");

    expect(mockFn).toBeCalledWith("something");
    expect(mockFn).toBeCalledWith("");
  });

  it(`SearchBar renders correctly`, async () => {
    const tree = renderer.create(<SearchBar onSearch={() => {}} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
