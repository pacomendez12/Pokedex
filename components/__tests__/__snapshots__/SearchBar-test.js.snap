// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar> tests SearchBar renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "flexDirection": "row",
        "height": 59,
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingTop": 10,
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "flexGrow": 1,
        "paddingHorizontal": 10,
        "width": 100,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f3f3f3",
          "borderBottomWidth": 0,
          "borderColor": "#86939e",
          "borderRadius": 10,
          "flexDirection": "row",
          "height": 40,
          "paddingLeft": 10,
          "paddingRight": 10,
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#f3f3f3",
            "height": 40,
            "justifyContent": "center",
            "marginVertical": 4,
            "paddingRight": 4,
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#8f8f91",
                      "fontSize": 24,
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                    Object {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
                testID="iconIcon"
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
      <TextInput
        allowFontScaling={true}
        editable={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Search"
        placeholderTextColor="#86939e"
        rejectResponderTermination={true}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "#f3f3f3",
            "color": "#000",
            "flex": 1,
            "fontSize": 18,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#f3f3f3",
            "height": 40,
            "justifyContent": "center",
            "marginVertical": 4,
            "paddingRight": 4,
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#8f8f91",
                      "fontSize": 24,
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                    Object {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
                testID="iconIcon"
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "#ff190c",
          "fontSize": 12,
          "margin": 5,
        }
      }
    />
  </View>
</View>
`;
